[Refactor] 보행등 코드 커버리지 높이기

ByungilOh-Fillip opened the pull request:
## #️⃣ 연관된 이슈
#215 


## ✅ 체크리스트

- [x] 🔀 PR 제목의 형식 맞추기 <!-- ex) [Feature] ㅇㅇ 기능 추가 -->
- [x] 💡 이슈 등록
- [x] 🏷️ 라벨 등록
- [x] 🧹 불필요한 코드 제거
- [x] 🧪 로컬 테스트 완료
- [x] 🏗️ 빌드 성공
- [x] 💯 테스트 통과


## 📝 작업 내용
<!-- 이번 PR에서 작업한 내용을 설명해 주세요. (이미지 첨부 가능) -->

Traffic 도메인 log 처리 추가
Traffic 도메인 Jacoco 커버리지 달성
Traffic 도메인 refactor 진행 및 작업 추가 사항 생성
- service에서 repository를 직접적으로 접근해서 사용하지 않게 수정
- 역할 분리가 명확하지 않아 테스트 시 어려움이 있어 작업 필요


## 👀 리뷰어 가이드라인
<!-- 리뷰어가 중점적으로 확인해야 할 사항을 작성해 주세요. -->
<!-- ex) 메서드 XXX의 이름을 더 잘 짓고 싶은데 혹시 좋은 명칭이 있을까요? -->

로그랑 테스트 코드 커버리지 맞게 올렸는데 부족한 부분이나 내용 추가할 사항 있다면 말씀해주세요!
TrafficService TODO 내용도 보시고 알려주고 싶은 부분있으시면 편하게 부탁드립니다!

ByungilOh-Fillip committed:
test: CustomTrafficRepoImpl 테스트 작성

ByungilOh-Fillip committed:
test: TrafficController 테스트 수정

ByungilOh-Fillip committed:
test: TrafficCsvService 테스트 수정 및 작성

ByungilOh-Fillip committed:
refactor: Traffic 도메인 로그 추가 및 코드 리펙토링

ByungilOh-Fillip committed:
test: TrafficRedisRepository 테스트 작성

ByungilOh-Fillip committed:
test: TrafficService 테스트 작성

++ refactor 사항 추가
crossroad 리펙토링 진행하며 작업 진행

ByungilOh-Fillip committed:
docs: 테스트 진행하며 더미 데이터 재 정의

DongminL commented on src/main/java/org/programmers/signalbuddyfinal/domain/trafficSignal/service/TrafficService.java:
```diff
@@ -22,53 +22,69 @@
-            return trafficRedisRepository.findNearbyTraffics(lat, lng, kiloRadius);
+            List<TrafficResponse> responseRedis = trafficRedisRepository.findNearbyTraffics(lat, lng, kiloRadius);
+
+            log.info("redis 주변 보행등 데이터 : redis data 갯수 = {} ", responseRedis.size());
+            return responseRedis;
         }
 
         try {
             responseDB = customTrafficRepository.findNearestTraffics(lat, lng, radius);
 
+            log.info("주변 보행등 정보 캐싱 : DB data 갯수 = {} ", responseDB.size());
             for (TrafficResponse response : responseDB) {
                 trafficRedisRepository.save(response);
             }
 
+            log.info("DB 주변 보행등 데이터 캐싱 성공");
```
여기에 있는 로그 레벨은 info에서 debug로 내려도 될 거 같아요

ByungilOh-Fillip replied:
엇 그러네여! 감사합니다!

DongminL commented on src/main/java/org/programmers/signalbuddyfinal/domain/trafficSignal/repository/TrafficRedisRepository.java:
```diff
@@ -38,78 +40,91 @@ public TrafficRedisRepository(RedisTemplate<Object,Object> redisTemplate){
+        trafficData.put("district", trafficResponse.getDistrict());
+        trafficData.put("signalType", trafficResponse.getSignalType());
+        trafficData.put("address", trafficResponse.getAddress());
 
-            // HASH 데이터 저장
-            Map<String, String> trafficData = new HashMap<>();
-            trafficData.put("serialNumber", String.valueOf(trafficResponse.getSerialNumber()));
-            trafficData.put("district", trafficResponse.getDistrict());
-            trafficData.put("signalType", trafficResponse.getSignalType());
-            trafficData.put("address", trafficResponse.getAddress());
+        hashOperations.put(KEY_HASH, trafficId.toString(), trafficData);
```
``` java
public void putStateCache(Long crossroadId, CrossroadStateResponse response) {
        ValueOperations<Object, Object> operations = redisTemplate.opsForValue();

        int minTimeLeft = response.minTimeLeft();
        minTimeLeft *= 100; // 1/10초 단위를 1/1000(ms)로 변환

        operations.set(STATE_PREFIX + crossroadId, response, minTimeLeft, TimeUnit.MILLISECONDS);
    }
```

이 코드처럼 DTO 자체를 직렬화 시켜서 저장시킬 수 있습니다. 이때 저장하려는 DTO는 `Serializable`을 상속받으면 돼요!

ByungilOh-Fillip replied:
오...변경해서 다시올려볼게요! 감사합니다!

DongminL commented on src/main/java/org/programmers/signalbuddyfinal/domain/trafficSignal/service/TrafficService.java:
```diff
@@ -22,53 +22,69 @@
 public class TrafficService {
 
+    private static final String TRAFFIC_REDIS_KEY = "traffic:info";
+
     private final CustomTrafficRepositoryImpl customTrafficRepository;
     private final TrafficRedisRepository trafficRedisRepository;
     private final TrafficRepository trafficRepository;
     private final RedisTemplate<Object, Object> redisTemplate;
 
     public List<TrafficResponse> searchAndSaveTraffic(Double lat, Double lng, int radius){
 
+        log.debug("주변 보행등 정보 - lat = {}, lng = {}, radius = {}", lat, lng, radius);
         List<TrafficResponse> responseDB;
 
-        boolean exists = Boolean.TRUE.equals(redisTemplate.hasKey("traffic:info"));
+        boolean exists = Boolean.TRUE.equals(redisTemplate.hasKey(TRAFFIC_REDIS_KEY));
```
`TrafficRedisRepository`로 Redis 로직을 분리하셨는데, 다시 `TrafficService`에 Redis 로직을 넣는 건 분리한 의미가 사라지는 것 같아요!

ByungilOh-Fillip replied:
그러네요.. 직접 접근을 두번하는 모양이되겠네요..! 이것도 수정해두겠습니다!

ByungilOh-Fillip commented:
## 리뷰 수정사항
- [x] service log info->debug로 수정
- [x] 불필요한 Map 객체 이용 -> dto로 바로 받게 변경
- [x] repository와 service에서 redis 중복 호출 -> repository에서만 호출하도록 분리

### 변경으로 생긴 테스트 취약 및 통과하지 못하는 테스트 수정
- [x] [TrafficRedisRepositoryTest] - 주변 보행등 redis 데이터 반환
- [x] [TrafficRedisRepositoryTest] - 보행등 저장 테스트
- [x] [TrafficServiceTest] - 주변 보행등 정보 redis 테스트

ByungilOh-Fillip committed:
refactor: pr 코드리뷰 적용

## 리뷰 수정사항
- service log info-> debug로 수정
- 불필요한 Map 객체 이용 -> dto로 바로 받게 변경
- repository와 service에서 redis 중복 호출 -> repository에서만 호출하도록 분리
- 테스트 코드 수정 -> 변경된 내용으로 mock 수정

zzuharchive commented on src/test/java/org/programmers/signalbuddyfinal/domain/trafficSignal/repository/CustomTrafficRepoImplTest.java:
```diff
@@ -0,0 +1,69 @@
+public class CustomTrafficRepoImplTest extends RepositoryTest {
+
+    @Mock
+    private CustomTrafficRepository customTrafficRepository;
```
Repository 테스트는 실제로 쿼리동작을 검증하는 목적으로 작성되기 때문에, mock보다는 **실제 Repository를 @ Autowired해서 테스트**하는 방식이 더 적절할 것 같습니다!
mock을 사용하면 when-then으로 지정한 값이 그대로 리턴되기 때문에, 쿼리의 실제 작동 여부를 확인하기는 어려울 수 있을 것 같아요!

ByungilOh-Fillip replied:
엇 그러게요..! 감사합니다! 테스트 수정해서 올려둘게요!

DongminL commented on src/main/java/org/programmers/signalbuddyfinal/domain/trafficSignal/service/TrafficService.java:
```diff
@@ -22,53 +20,66 @@
 @Transactional(readOnly = true)
 public class TrafficService {
 
+    private static final String TRAFFIC_REDIS_KEY = "traffic:info";
```
사용하지 않는 변수는 이제 필요없으니 삭제하면 좋을 거 같아요

ByungilOh-Fillip replied:
헉 그러게요..! 바로 삭제할게요!

DongminL commented on src/test/java/org/programmers/signalbuddyfinal/domain/trafficSignal/service/TrafficServiceTest.java:
```diff
@@ -0,0 +1,129 @@
+
+    }
+
+    /**
+     * TODO: TrafficService 코드 refactor 필요
+     *   -> findById()
+     *      Servive 레이어에서 entity에 접근하는 repository를 직접 사용하지 않게
+     */
+/*    @Test
+    @DisplayName("보행등 ID 캐싱 테스트")
+    void testTrafficFindByIdRedisNotExists() {
+
+        // Given
+        Long id = expected.get(0).getTrafficSignalId();
+
+        when(redisTemplate.hasKey(TRAFFIC_REDIS_KEY)).thenReturn(false);
+        when(trafficRepository.findById(id)).thenReturn(expected.get(0));
+
+        // When
+        TrafficResponse result = trafficService.trafficFindById(id);
+
+        // Then
+        assertThat(result).isNotNull();
+    }*/
+}
```
이 테스트는 왜 주석 처리를 하셨는지 알 수 있을까요?

ByungilOh-Fillip replied:
레포지토리에서 반환된 엔티티를

> 이 테스트는 왜 주석 처리를 하셨는지 알 수 있을까요?

서비스에서 repository.findbyid()를 이용하면서 테스트에서 엔티티를 모킹해서 사용하는 경우가 생겼는데
엔티티를 모킹해서 사용하면 JPA에서 영속성 컨텍스트 문제나 프록시 충돌같은 문제가 생길 수 있다고 해서 dto객체를 이용하는걸로 서비스를 리펙토링 해야겠다고 판단했었어요! 

그런데 테스트 코드 작업을 어느정도 해둔 상태였어서 리펙토링 작업을 새로운 이슈생성해서 진행하고 완료 후 테스트 코드도 다시 정리해서 올리려구 남겨뒀습니다! 

따로 메모해둬도 되는데 우선 지워둘까요?

DongminL replied:
기존 코드가 JPA를 사용해서 `findById`로 Entity를 가져왔고, DTO로 변환하여 사용하지 않았다는 말씀이신거죠?

ByungilOh-Fillip replied:
네네 맞습니다

DongminL replied:
아하 그럼 리팩터링 하신 다음에 테스트 해보시면 될 것 같습니다!

ByungilOh-Fillip replied:
네 답변 감사합니다! ㅎㅎ

DongminL commented on src/main/java/org/programmers/signalbuddyfinal/domain/trafficSignal/repository/TrafficRedisRepository.java:
```diff
@@ -50,13 +52,7 @@ public void save(TrafficResponse trafficResponse) {
-        trafficData.put("address", trafficResponse.getAddress());
-
-        hashOperations.put(KEY_HASH, trafficId.toString(), trafficData);
+        hashOperations.put(KEY_HASH, trafficId.toString(), trafficResponse);
```
Redis에 Hash 구조로 저장된 데이터를 삭제할 때, 관련 데이터가 전부 삭제되는 것을 확인하셨을까요?

그리고 `ValueOperations` 를 사용하시면 객체가 통째로 직렬화가 되어 저장됩니다. 역직렬화도 마찬가지로 가능합니다.

ByungilOh-Fillip replied:
삭제할 때가 TTL이 모두 소진됐을 때를 말씀하시는거 맞죠? 그럼 삭제는 잘 됩니다! 
ValueOperations 그래서 쓰는거군여..! 저도 그럼 코드 변경해볼게요! 감사합니다 ㅎㅎ

ByungilOh-Fillip commented:
## 리뷰 사항 체크리스트
- [x] ValueOperation 적용
- [x] Repository 테스트 수정

ByungilOh-Fillip committed:
refactor: pr 코드리뷰 적용

## 리뷰 수정사항
- TrafficRedisRepository hash->value로 변경
- TrafficService 불필요한 코드제거
- CustomTrafficRepoImplTest
    - RepositoryTest로 변경

ByungilOh-Fillip committed:
fix: build error 해결

원인
KEY_INFO가 Redis 예약어로 설정되어있어 충돌이 나는 문제

해결
KEY_INFO->KEY_VALUE로 수정
